### PR TITLE
add support for user defined encoding & decoding

### PR DESCRIPTION
**Pull request by fmoor**
fixes https://github.com/edgedb/edgedb-go/issues/47
fixes https://github.com/edgedb/edgedb-go/issues/114

The following changes were made:

- create a distinct descriptor type. Reading a descriptor off the wire goes form leaves to root of the type tree. Applying a go type to a descriptor to create a codec goes from root to leaves. Having a distinct descriptor type makes it easier to move root to leaves through both type trees at the same time to build a codec.
- split codecs into Encoders and Decoders to enable separate Marshaler and Unmarshaler interfaces.
- move scalar codecs to smaller files to keep file sizes reasonable.
- remove logic for skipping \_\_tid__ fields
- add support for Marshler and Unmarshaler interfaces which allow custom encoding and decoding for all scalar types except arrays. This API was inspired by the stdlib https://golang.org/pkg/encoding/#BinaryMarshaler
- test Marshaler and Unmarshaler for all scalar types.

# Custom Codec Example
```go
type DecimalUnmarshaler struct {
    // a real implementation would probably do more than store the raw bytes off the wire
    data []byte
}

func (d *DecimalUnmarshaler) UnmarshalEdgeDBDecimal(data []byte) error {
    // decode data from the server
    d.data = make([]byte, len(data))
    copy(d.data, data)
    return nil
}

func main() {
    var result DecimalMarshaler
    err := con.Query(ctx, "SELECT 1234n", &result)
    if err != nil {
        log.Fatal(err)
    }

    fmt.Println(result)
}
```

### CODE REVIEW

File: types.go [hunk 1]
```diff
@@ -33,6 +36,110 @@ type (
...
+	BoolMarshaler = codecs.BoolMarshaler
+
+	// BoolUnmarshaler enables decoding user defined bool values.
+	BoolUnmarshaler = codecs.BoolUnmarshaler
+
+	// BytesMarshaler enables encoding user defined bytes values.
+	BytesMarshaler = codecs.BytesMarshaler
+
+	// BytesUnmarshaler enables decoding user defined bytes values.
+	BytesUnmarshaler = codecs.BytesUnmarshaler
```
Comment (fmoor) [inline types.go:50]: This long list of types clutters the documentation for the edgedb package. These types don't need to be public. User code doesn't need access to them. They just need to be documented. It is probably better either:

1. to leave them in `internal/codecs`, don't make them part of the public API, and reference them in the introduction to the edgedb package telling users that they can implement them.
2. create a second public package that only has these interfaces. Maybe `edgedbmarshalers` or something. This way it is obvious that they are part of the public API but it keeps the documentation cleaner.

Comment (1st1) [reply types.go:50]: > to leave them in internal/codecs, don't make them part of the public API, and reference them in the introduction to the edgedb package telling users that they can implement them.

I'd go with this option.

**Comment by 1st1** [comment]
Is there any performance degradation caused by the new, more advanced, marshal/unmarshal code?

**Comment by 1st1** [comment]
The type name in this example is a bit confusing:

```go
type DecimalUnmarshaler struct {
    // a real implementation would probably do more than store the raw bytes off the wire
    data []byte
}
```

IMO it would be more readable if the struct was named `CustomDecimal`. Same applies to unittests/docs.

**Review by 1st1 — APPROVED** [review]
Solid work and everything looks good and well thought out. Before merging this though, I'm wondering what's the impact on performance looks like. We now switch based on `val.(type)` in every marshaller.

**Comment by fmoor** [comment]
> I'm wondering what's the impact on performance looks like. We now switch based on `val.(type)` in every marshaller.

I'll run the benchmarks on it. I would be if this change made a significant difference. That type switch used to be a type assertion which is a very similar mechanism.

File: types.go (hunk 1 above)
Comment (fmoor) [reply types.go]: changed

**Comment by fmoor** [comment]
It doesn't seem like there is a change in performance with this PR.

# Before
image

# After
image

**Pull request merged**